feat: add section 4 of report [RWDQA-34]

tomzemp opened the pull request:
This PR adds in Section 4 of the annual report (https://dhis2.atlassian.net/browse/RWDQA-34). 

For calculation explanations see: https://docs.google.com/document/d/1_fO30sLZk2LYKuUh1Toqi3GYW8kVPGwjj5e0bv8HnZw/edit#bookmark=id.tvitxy5hwd9b
_(This section replicates what the old app is doing and discrepancies were not noted while testing against the Maternal Health group on demos.dhis2.org/dq)_

This PR:

- adds denominator relations into the mappedConfiguration (see src/utils/utils.js)
- adds a custom hook (useFetchSectionFourData) for retrieving Section 4 data. It's in a custom hook because the org unit LEVEL can vary based on the denominator relation (this replicates what the old app is doing, though some of the actual decision making for how this level is determined does not make a lot of sense to me and is not documented in the user guide, so we might potentially follow up with Olav on this and see if we can simplify to just make requests at the selected LEVEL (where possible)
- adds a SectionFour component (which refetches data when reportParameters updates. Note that the intent is that these are memoized to prevent refetch unless something with selections has changed). This is not styled (handled as part of https://dhis2.atlassian.net/browse/RWDQA-41)

tomzemp commented on src/components/annual-report/report-data/utils/utils.js:
```diff
@@ -0,0 +1,35 @@
+    return rowData
+}
+
+export const mapSectionFourResponse = ({ headers, rows }) => {
```
this maps to a `{[dx]: {[ou]: {[pe]: [value]}}}` structure. 

Here the periods should all be current period and for the overall response there should only be one organisation unit, but I thought for consistency purposes, it's probably to keep data in one structure and just provide the period/orgUnit for retrieval from this data structure.

KaiVandivier commented on src/components/annual-report/report-data/section4/useFetchSectionFourData.js:
```diff
@@ -0,0 +1,111 @@
+    const engine = useDataEngine()
+
+    const refetch = async ({ variables = {} }) => {
+        const { denominatorRelations } = variables?.mappedConfiguration
+        const denominatorRelationDEs = [
+            ...denominatorRelations?.map((rel) => rel.A.id),
+            ...denominatorRelations?.map((rel) => rel.B.id),
+        ]
```
My linter flagged these lines with "Unsafe usage of optional chaining. If it short-circuits with 'undefined' the evaluation will throw TypeError" -- here's a suggestion:
```suggestion
        const denominatorRelations = variables?.mappedConfiguration?.denominatorRelations
        const denominatorRelationDEs = denominatorRelations ? [
            ...denominatorRelations.map((rel) => rel.A.id),
            ...denominatorRelations.map((rel) => rel.B.id),
        ] : []
```

tomzemp replied:
Ah. Yeah. I see that optional chaining is kind of useless when rereading it 😄.

I guess it would make sense to address it even earlier in the code where these denominatorRelations get summarized into mappedConfiguration (https://github.com/hisprwanda/who-data-quality-annual-report/blob/933933c10bfeb54ffc9d25213d56c908194fb47a/src/utils/utils.js#L75-L76). Will also update that

KaiVandivier commented on src/components/annual-report/report-data/section4/useFetchSectionFourData.js:
```diff
@@ -0,0 +1,111 @@
+
+    const engine = useDataEngine()
+
+    const refetch = async ({ variables = {} }) => {
```
I saw a comment in SectionFour.js that you want this to be stable -- you could wrap this in `useCallback` with `engine` as a dependency:
```js
const refetch = useCallback(async ({ variables = {} }) => {
    ...
}, [engine])
```

KaiVandivier commented on src/components/annual-report/report-data/section4/useFetchSectionFourData.js:
```diff
@@ -0,0 +1,111 @@
+        // we need to preserve information on index to tie response back to relevant denominator relation (some can be filtered out)
+        const byLevelRequestsWithIndices = denominatorRelations
+            .map((rel, index) => {
+                // UN type does not require details by subOrgUnit level, so these can be skipped
+                if (rel.type === 'un') {
+                    return
+                }
+
+                // the level will be the minimum of the lowest of the denominator levels and the selected level
+                const denomMinLevel = Math.min(
+                    rel.A.lowLevel,
+                    rel.B.lowLevel,
+                    variables.orgUnitLevelNumber
+                )
+                // if the denominator relation level is less than the boundary org unit's level, it cannot be calculated
+                if (denomMinLevel < variables.boundaryOrgUnitLevel) {
+                    return
+                }
```
I saw in the calculations doc "if either denominator is missing a DE/indicator: skip the denom. rel. calculation" -- is that handled here? (sorry if I'm missing it 🙃)

tomzemp replied:
Yeah. That should be handled by filtering out denominator relations without ids (or levels) when we summarize from the configuration: https://github.com/hisprwanda/who-data-quality-annual-report/blob/933933c10bfeb54ffc9d25213d56c908194fb47a/src/utils/utils.js#L98

KaiVandivier replied:
Great thanks! 🙏

KaiVandivier commented on src/components/annual-report/report-data/section4/useFetchSectionFourData.js:
```diff
@@ -0,0 +1,111 @@
+            (r) => r.relation
+        )
+        try {
+            const dataByLevel = await Promise.all(byLevelRequests)
+            const overallData = await engine.query(section4Overall, {
+                variables: {
+                    dataElements: denominatorRelationDEs,
+                    orgUnits: variables.orgUnits,
+                    periods: [variables.currentPeriod.id],
+                },
+            })
```
Here's a minor optimization that could allow the requests to happen in parallel:

```suggestion
            const overallRequest = engine.query(section4Overall, {
                variables: {
                    dataElements: denominatorRelationDEs,
                    orgUnits: variables.orgUnits,
                    periods: [variables.currentPeriod.id],
                },
            })
            const [overallData, ...dataByLevel] = await Promise.all([
                overallRequest,
                byLevelRequests,
            ])
```

KaiVandivier commented on src/components/annual-report/report-data/utils/mathService.js:
```diff
@@ -0,0 +1,7 @@
+export const getRoundedValue = (val, places) => {
+    if (isNaN(Number(val))) {
+        return val
+    }
+
+    return Math.round(val * Math.pow(10, places)) / Math.pow(10, places)
+}
```
This is an elegant solution to the rounding problem 🙂 There's [`<number>.toFixed(places)`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toFixed) available -- would that do what you need to do here?

tomzemp replied:
Yeah. `toFixed` would be easier. The current app rounds*, so for example 25.123 is rounded to 25.1, but 25 remains 25. `Number(25).toFixed(1)` though would result in `25.1`. I don't know if this is an important decision, so maybe it makes sense to change to avoid using our own rounding function?

* https://github.com/dhis2/who-data-quality-app/blob/master/src/appCore/appServiceMath.js#L183-L189

KaiVandivier replied:
Ah good point -- I think we should keep `getRoundedValue`; I think it's better and a bit more flexible

KaiVandivier commented on src/components/annual-report/report-data/section4/section4Calculations.js:
```diff
@@ -0,0 +1,158 @@
+        overallScore,
+        name: relation.name,
+        A: metadata[relation.A.id]?.name,
+        B: metadata[relation.A.id]?.name,
```
```suggestion
        B: metadata[relation.B.id]?.name,
```

tomzemp replied:
thanks!

KaiVandivier commented on src/components/annual-report/report-data/section4/SectionFour.js:
```diff
@@ -0,0 +1,175 @@
+    const { loading, data, error, refetch } = useFetchSectionFourData()
+
+    useEffect(() => {
+        // this is hardcoded until we address period selection
+        const currentPeriod = {
+            id: '2022',
+            startDate: '2022-01-01',
+            endDate: '2022-12-31',
+        }
```
Just a reminder to update this 🙂

KaiVandivier commented on src/components/annual-report/report-data/section4/useFetchSectionFourData.js:
```diff
@@ -0,0 +1,111 @@
+import { useState } from 'react'
+
+const section4ByLevel = {
+    data_detail_by_level: {
```
It would be great to use the constants you have set up in `section4Calculations.js` like `const OVERALL_RESPONSE_NAME = 'data_over_all_org_units'` as the keys here -- maybe define and export them in this file and import them in `section4Calculations.js`

KaiVandivier commented on src/components/annual-report/report-data/section4/section4Calculations.js:
```diff
@@ -0,0 +1,158 @@
+}) => {
+    const section4a = []
+    for (const rel of denominatorRelations) {
+        if (rel.type !== 'un') {
+            break
+        }
```
Are `un`-type relations necessarily first in the array? If not, they might be missed -- using `continue` here instead of `break` would fix that

Not related to the code here, but I'm also curious -- is there ever more than 1 `un`-type relation? 🤔

tomzemp replied:
Yikes 🙈. That should be `continue`. Thanks!

KaiVandivier commented on src/components/annual-report/report-data/section4/section4Calculations.js:
```diff
@@ -0,0 +1,158 @@
+            break
+        }
+        section4a.push({
+            name: rel.name,
```
Does this particular section get a hard-coded name? I see "Ratio of population projection from the Bureau of Statistics to a UN projection" in the old app, but maybe that's a different object

tomzemp replied:
Yeah...this is kind of weird in the old app, but it let you define multiple denominator relations of type `un` and then only used the last one in the report and put it with that hardcoded name.

I spoke with Olav, and we decided to let you add multiple ones in the configuration, and have them all displayed in the report (hence we cannot use the hard-coded name).

(See slack conversation: https://dhis2.slack.com/archives/C053JT69YD8/p1696500142931519)

KaiVandivier reviewed (commented):
Nice work! Compared it to the old app and the values agree, and the calculations check out with the spec as best as I can tell

I noticed one small bug, had a few questions, and shared a few suggestions 🙂

tomzemp commented:
Thanks for the review @KaiVandivier 🙏! I've implemented your suggested commits, fixed the issues you identified, and left some comments.

I'll check the PR for Section 3 and will update that PR with any points that would apply from this review

KaiVandivier reviewed (approved):
Nice, looks great! 🥳 🙌

KaiVandivier commented on src/components/annual-report/report-data/section4/useFetchSectionFourData.js:
```diff
@@ -0,0 +1,128 @@
+
+                setData({
+                    ...overallData,
+                    data_detail_by_level: dataByLevel.map(
+                        (resp) => resp?.data_detail_by_level
+                    ),
```
Ah actually just noticed this (while looking at section 3):
```suggestion
                    [BY_LEVEL_RESPONSE_NAME]: dataByLevel.map(
                        (resp) => resp?.[BY_LEVEL_RESPONSE_NAME]
                    ),
```
